src/droidplug/peripheral.rs

```diff
@@ -280,18 +280,18 @@ impl api::Peripheral for Peripheral {
                             characteristic_uuid: characteristic.get_uuid()?,
                         });
                     }
-                    characteristics.insert(Characteristic {
+                    let char = Characteristic {
                         service_uuid: service.get_uuid()?,
                         uuid: characteristic.get_uuid()?,
                         properties: characteristic.get_properties()?,
                         descriptors: descriptors.clone(),
-                    });
-                    peripheral_characteristics.push(Characteristic {
-                        service_uuid: service.get_uuid()?,
-                        uuid: characteristic.get_uuid()?,
-                        properties: characteristic.get_properties()?,
-                        descriptors: descriptors,
-                    });
+                    };
+                    // Only consider the first characteristic of each UUID
+                    // This "should" be unique, but of course it's not enforced
+                    if !characteristics.contains(&char) {
```

Comment:
What if there is another characteristic which has the same UUID but is different in some other way?

Reply:
It's dropped. Ideally, this should never happen, but the reality turns out that there are devices out there that do this.

Right now, the first characteristic is overridden, so in that respect its no worse (this change just picks the first one rather than the last).

I did consider changing the characteristic structures to Map<Uuid, Vec<CharacteristicObj>> (please excuse the pseudocode), but given that all the other methods use the Uuid as the key for performing characteristic operations, that'd mean that they'd all have to be able to take a UUID and an index (99.9% of the time, there's only going to be 1 item in the array). The trade-off between being able to address all characteristics (even with duplicate keys) vs the overhead of the interface changes to support 1 device felt way off to me.

Reply:
How? This code seems to be comparing the characteristic as a whole, not just the UUID, unlike the other platform implementations.

Reply:
That's a good question... Sorry, I was answering for the change as a whole, not the Android diff specifically.

Hmm... the comparator does look like it applies to all members of the struct, so the contains check probably never returns true. I guess the properties and then descriptors would define the order that btleplug presents the characteristics, and I'm guessing that at a higher level they'll get uniqued by UUID. I see the UUID is the only identifier passed across the JNI boundary, and the characteristics are held in an ArrayList on that side, so I think making this set UUID unique (first characteristic wins) will mean that the presented properties match up the the characteristic being addressed.

Reply:
@qwandor I've re-written and re-tested the Android portion of the change